nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-services/src/main/java/org/apache/nifi/mongodb/MongoDBLookupService.java

```diff
@@ -52,68 +54,125 @@
+                return Optional.ofNullable(new MapRecord(toUse, result));
             }
         } catch (Exception ex) {
             getLogger().error("Error during lookup {}", new Object[]{ query.toJson() }, ex);
             throw new LookupFailureException(ex);
         }
     }
 
-    @Override
+    private RecordSchema loadSchema(Map<String, Object> coordinates) {
+        Map<String, String> variables = coordinates.entrySet().stream()
+            .collect(Collectors.toMap(
+                e -> e.getKey(),
+                e -> e.getValue().toString()
+            ));
+        try {
+            return getSchema(variables, null);
+        } catch (Exception ex) {
+            return null;
+        }
+    }
+
+    private RecordSchema convertSchema(Map<String, Object> result) {
+        List<RecordField> fields = new ArrayList<>();
+        for (Map.Entry<String, Object> entry : result.entrySet()) {
+
+            RecordField field;
+            if (entry.getValue() instanceof Integer) {
+                field = new RecordField(entry.getKey(), RecordFieldType.INT.getDataType());
+            } else if (entry.getValue() instanceof Long) {
+                field = new RecordField(entry.getKey(), RecordFieldType.LONG.getDataType());
+            } else if (entry.getValue() instanceof Boolean) {
+                field = new RecordField(entry.getKey(), RecordFieldType.BOOLEAN.getDataType());
+            } else if (entry.getValue() instanceof Double) {
+                field = new RecordField(entry.getKey(), RecordFieldType.DOUBLE.getDataType());
+            } else if (entry.getValue() instanceof Date) {
+                field = new RecordField(entry.getKey(), RecordFieldType.DATE.getDataType());
+            } else if (entry.getValue() instanceof List) {
+                field = new RecordField(entry.getKey(), RecordFieldType.ARRAY.getDataType());
+            } else if (entry.getValue() instanceof Map) {
+                RecordSchema nestedSchema = convertSchema((Map)entry.getValue());
+                RecordDataType rdt = new RecordDataType(nestedSchema);
+                field = new RecordField(entry.getKey(), rdt);
+            } else {
+                field = new RecordField(entry.getKey(), RecordFieldType.STRING.getDataType());
+            }
+            fields.add(field);
+        }
+
+        return new SimpleRecordSchema(fields);
+    }
+
+    private volatile Document projection;
+    private MongoDBControllerService controllerService;
+    private String schemaNameProperty;
+
     @OnEnabled
-    public void onEnabled(final ConfigurationContext context) throws InitializationException, IOException, InterruptedException {
+    public void onEnabled(final ConfigurationContext context) {
         this.lookupValueField = context.getProperty(LOOKUP_VALUE_FIELD).getValue();
-        super.onEnabled(context);
+        this.controllerService = context.getProperty(CONTROLLER_SERVICE).asControllerService(MongoDBControllerService.class);
```

Comment:
I get a runtime error here as it is expecting an interface not an impl class, I had to change this to `MongoDBClientService` to get past the runtime error.

Reply:
Looks like this one is still there, still getting the runtime errors